src/middlewares/languageMiddleware.ts

```diff
@@ -0,0 +1,169 @@
+import Negotiator from "negotiator";
+import { headers } from "next/headers";
+import { NextRequest, NextResponse } from "next/server";
+
+import { client } from "studio/lib/client";
+import { SlugTranslations } from "studio/lib/interfaces/slugTranslations";
+import { LanguageObject } from "studio/lib/interfaces/supportedLanguages";
+import {
+  DEFAULT_LANGUAGE_QUERY,
+  LANGUAGES_QUERY,
+} from "studio/lib/queries/languages";
+import {
+  SLUG_TRANSLATIONS_FROM_LANGUAGE_QUERY,
+  SLUG_TRANSLATIONS_TO_LANGUAGE_QUERY,
+} from "studio/lib/queries/slugTranslations";
+
+async function translateSlug(
+  slug: string,
+  targetLanguageId: string,
+  sourceLanguageId?: string,
+): Promise<string | undefined> {
+  if (slug.length === 0) {
+    return slug;
+  }
+  const slugTranslations = await client.fetch<SlugTranslations | null>(
+    sourceLanguageId !== undefined
+      ? SLUG_TRANSLATIONS_FROM_LANGUAGE_QUERY
+      : SLUG_TRANSLATIONS_TO_LANGUAGE_QUERY,
+    {
+      slug,
+      language: sourceLanguageId ?? targetLanguageId,
+    },
+  );
+  return slugTranslations?._translations.find(
+    (t) => t !== null && t.language === targetLanguageId,
```

Comment:
like mentioned before, not a fan of single letter parameter
```suggestion
    (translation) => translation !== null && translation.language === targetLanguageId,
```